### PR TITLE
Removed Vehicle Entity and added custom exception hanlder

### PR DESCRIPTION
**Pull request by karanbadhwar**

### CODE REVIEW

**Comment by karanbadhwar** [comment]
Had to remove appplication.properties, got error and emails regarding secret info shared

File: src/main/java/com/volvo/tax_calculator/controller/TaxCalculate.java [hunk 1]
```diff
@@ -39,66 +32,68 @@ public String healthCheck() {
...
+            }
             ResponseDto<Integer> response = calculateTaxForVehicle(vehicle);
             return ResponseEntity.ok(response);
-        } catch (Exception e) {
-            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(new ResponseDto<>(HttpStatus.INTERNAL_SERVER_ERROR, "Error calculating tax", null));
+        } catch (VehicleDataMissingException | TaxCalculationException ex ) {
+            throw ex;
+        } catch(Exception ex) {
+            log.error("Unexpected error during tax calculation", ex);
+            throw new TaxCalculationException("Error calculating tax");
```
Comment (guptanshum) [inline src/main/java/com/volvo/tax_calculator/controller/TaxCalculate.java:46]: Don't catch these

File: src/main/java/com/volvo/tax_calculator/controller/TaxCalculate.java [hunk 2]
```diff
@@ -39,66 +32,68 @@ public String healthCheck() {
...
         }
     }
 
     public ResponseDto<Integer> calculateTaxForVehicle(VehicleEntity vehicle) {
         if (vehicle == null) {
-            log.error("Vehicle entity is null");
-            return new ResponseDto<>(HttpStatus.BAD_REQUEST, "Vehicle data is missing", 0);
+            throw new VehicleDataMissingException("Vehicle data is missing");
```
Comment (guptanshum) [inline src/main/java/com/volvo/tax_calculator/controller/TaxCalculate.java:52]: Can this be null here?

File: src/main/java/com/volvo/tax_calculator/controller/TaxCalculate.java [hunk 3]
```diff
@@ -39,66 +32,68 @@ public String healthCheck() {
...
-        String vehicleType = vehicle.getVehicleType();
 
-        if (isExempted(vehicleType)) {
+        if (isExempted(vehicle.vehicleType())) {
             return new ResponseDto<>(HttpStatus.OK, "Vehicle is exempted from tax", 0);
         }
 
         if (taxCalculatorService.isInNonTaxablePeriod()) {
             return new ResponseDto<>(HttpStatus.OK, "Currently in Non-taxable period", 0);
         }
```
Comment (guptanshum) [inline src/main/java/com/volvo/tax_calculator/controller/TaxCalculate.java:61]: Extract return statements to a method

File: src/main/java/com/volvo/tax_calculator/controller/TaxCalculate.java [hunk 4]
```diff
@@ -39,66 +32,68 @@ public String healthCheck() {
...
-            } else {
-                return new ResponseDto<>(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to create tax entry", null);
-            }
-        } else {
-            // Handle tax update based on previous tax entry
-            return calculateOrUpdateTaxForExistingRecord(currentVehicle, currentHourTax);
-
+        TaxEntity lastRecentTax = taxCalculatorService.getLatestTax(vehicle.vehicleNumber());
+        LocalDateTime currentTime = LocalDateTime.now();
+        if(lastRecentTax != null) {
```
Comment (guptanshum) [inline src/main/java/com/volvo/tax_calculator/controller/TaxCalculate.java:67]: Code is formatted

File: src/main/java/com/volvo/tax_calculator/service/TaxCalculatorService.java [hunk 5]
```diff
@@ -1,44 +1,32 @@
...
-    @Autowired
-    private VehicleRepository vehicleRepository;
-
-    @Autowired
-    private TaxRepository taxRepository;
-
-    @Value("${holidays}")  // Load the holidays from application.properties
-    private String holidays;
+    private final TaxRepository taxRepository;
+    private final Environment environment;
```
Comment (guptanshum) [inline src/main/java/com/volvo/tax_calculator/service/TaxCalculatorService.java:23]: This is not required

File: src/main/java/com/volvo/tax_calculator/service/TaxCalculatorService.java [hunk 6]
```diff
@@ -54,77 +42,89 @@ public boolean isInNonTaxablePeriod() {
         return isJuly || isWeekend || isHoliday;
     }
 
-    @Transactional
-    public boolean updatingTax(int currentTax, VehicleEntity vehicle, LocalDateTime currentTimeDate) {
+    public TaxEntity getLatestTax(String vehicleNumber) {
+        List<TaxEntity> taxList = taxRepository.findByVehicleNumberOrderByUpdatedAtDesc(vehicleNumber);
+
+        if(taxList.isEmpty()) {
```
Comment (guptanshum) [inline src/main/java/com/volvo/tax_calculator/service/TaxCalculatorService.java:48]: Code is not formatted

File: src/main/java/com/volvo/tax_calculator/service/TaxCalculatorService.java [hunk 7]
```diff
@@ -54,77 +42,89 @@ public boolean isInNonTaxablePeriod() {
         return isJuly || isWeekend || isHoliday;
     }
 
-    @Transactional
-    public boolean updatingTax(int currentTax, VehicleEntity vehicle, LocalDateTime currentTimeDate) {
+    public TaxEntity getLatestTax(String vehicleNumber) {
```
Comment (guptanshum) [inline src/main/java/com/volvo/tax_calculator/service/TaxCalculatorService.java:45]: You can use Optional instead

**Pull request merged**